src/address.rs

```diff
@@ -68,66 +77,334 @@ impl Matcher {
+        tag("?").map(|_| AddressPatternComponent::WildcardSingle),
+        // Combinations of wildcards are a bit tricky.
+        // Multiple '*' wildcards in a row are equal to a single '*'.
+        // A '*' wildcard followed by any number of '?' wildcards is also equal to '*' but must
+        // match at least the same amount of characters as there are '?' wildcards in the combination.
+        // For example, '*??' must match at least 2 characters.
+        is_a("*?").map(|x: &str| AddressPatternComponent::Wildcard(x.matches('?').count())),
+        pattern_choice.map(|choices: Vec<&str>| {
+            AddressPatternComponent::Choice(choices.iter().map(|x| x.to_string()).collect())
+        }),
+        pattern_character_class
+            .map(|s: &str| AddressPatternComponent::CharacterClass(CharacterClass::new(s))),
+    ))(input)
 }
 
-fn map_question_mark(_: &str) -> String {
-    r"\w?".into()
+fn match_literally<'a>(input: &'a str, pattern: &str) -> IResult<&'a str, &'a str> {
+    tag(pattern)(input)
 }
 
-fn parse_address_part(input: &str) -> IResult<&str, &str> {
-    preceded(char('/'), take_till1(|c| " \t\r\n#*,/?[]{}".contains(c)))(input)
+fn match_wildcard_single(input: &str) -> IResult<&str, &str> {
+    // TODO: this has to be possible with a simpler parser?
+    verify(take(1usize), |s: &str| s.chars().all(is_address_character))(input)
 }
 
-fn parse_address_pattern_part(input: &str) -> IResult<&str, &str> {
-    preceded(char('/'), take_till1(|c| " \n\t\r/".contains(c)))(input)
+fn match_character_class<'a>(
+    input: &'a str,
+    character_class: &'a CharacterClass,
+) -> IResult<&'a str, &'a str> {
+    if character_class.negated {
+        is_not(character_class.characters.as_str())(input)
+    } else {
+        is_a(character_class.characters.as_str())(input)
+    }
 }
 
-// Translate OSC pattern rules into an regular expression.
-// A pattern part can contain more than one rule, e.g. `{voice,synth}-[1-9]` contains two rules, an alternative and a number range.
-fn parse_pattern_part(input: &str) -> IResult<&str, String> {
-    many1(alt((
-        delimited(char('{'), is_not("}"), char('}')).map(map_alternative),
-        delimited(tag("[!"), is_not("]"), char(']')).map(|s: &str| wrap_with(s, "[^", "]")),
-        delimited(char('['), is_not("]"), char(']')).map(|s: &str| wrap_with(s, "[", "]")),
-        tag("*").map(map_wildcard),
-        tag("?").map(map_question_mark),
-        is_not("[{").map(|s: &str| s.to_owned()),
+/// Try all the tags in a choice element
+/// Example choice element: '{foo,bar}'
+/// It will get parsed into a vector containing the strings "foo" and "bar", which are then matched
+fn match_choice<'a>(input: &'a str, choices: &[String]) -> IResult<&'a str, &'a str> {
+    for choice in choices {
+        if let Ok((i, o)) = tag::<_, _, nom::error::Error<&str>>(choice.as_str())(input) {
+            return Ok((i, o));
```

Comment:
The documentation says ` /// Try all the tags in a choice element` but won't this return after the first valid tag?

Reply:
I tried to clarify it

Reply:
That sounds clear to me 👍🏻